Update Me: 2FA backup codes list to have to ability to save codes to disk

noxee opened the pull request:
Issue #385

This pull request seeks to add a button to the backup codes list that will allow a user to save their codes to the disk. The filename for the saved codes is currently configured to `[USERNAME]-backup-codes.txt`.

**Implementation notes:**

This just uses the built-in browser `Blob` object and the `browser-filesaver` library to facilitate the saving of the file to disk.

**Testing instructions:**

1. Navigate to the two-step authentication page.
2. Click the Generate New Backup Codes button.
3. Once new codes have been generated click the Save button.
4. Confirm that the codes have been saved to disk successfully.

**Notes:**
While this currently users the built-in `Blob` object there is a limitation with this as there are some older browsers that don't implement it. If it is desired to support older browsers with this implementation the polyfill `blob-polyfill` could be used. I did notice that there is already a `blob` module installed but it looks like it only falls back to `BlobBuilder` if `Blob` cannot be found (and reading up on `BlobBuilder` it looks like it is not considered obsolete) so I'm not sure if that would be a viable option.

timmyc commented on client/me/security-2fa-backup-codes-list/index.jsx:
```diff
@@ -216,6 +227,17 @@ module.exports = React.createClass( {
+						onClick={ function( event ) {
+							analytics.ga.recordEvent( 'Me', 'Clicked On 2fa Save Backup Codes Button' );
+							this.saveCodesToFile( event );
+						}.bind( this ) }
```
to avoid the bind and inline function here, how about moving the analytics call to `saveCodesToFile`?

timmyc commented on client/me/security-2fa-backup-codes-list/index.jsx:
```diff
@@ -15,6 +15,8 @@ var	FormButton = require( 'components/forms/form-button' ),
 	config = require( 'config' ),
 	Notice = require( 'components/notice' );
 
+import { saveAs } from 'browser-filesaver';
```
Cool to see this being used elsewhere in the project 👍

timmyc commented on client/me/security-2fa-backup-codes-list/index.jsx:
```diff
@@ -61,6 +63,15 @@ module.exports = React.createClass( {
+	saveCodesToFile: function( event ) {
+		event.preventDefault();
+
+		const { user_login } = this.props.userSettings.settings;
```
I don't see this prop being used elsewhere on the page, I realize that although unlikely, do we need to guard against `userSettings` being undefined?

Another route would be to use [lib\user](https://github.com/Automattic/wp-calypso/tree/master/client/lib/user) to get at the login name.

timmyc commented on client/me/security-2fa-backup-codes-list/index.jsx:
```diff
@@ -216,6 +227,17 @@ module.exports = React.createClass( {
 					</FormButton>
 
 					<FormButton
+						className="security-2fa-backup-codes-list__save"
```
Since we have no styles for this className, I think this can be removed.

timmyc commented:
Thanks for this PR, this is a nice addition to the 2FA flow 👍 

I left a few notes above - and here is a screen shot of the new button for others:

<img width="724" alt="two-step_authentication_ _wordpress_com" src="https://cloud.githubusercontent.com/assets/22080/14693989/398bcc34-0719-11e6-9c99-a184f7f36ce0.png">

noxee commented:
@timmyc I've updated the code as per the suggestions you made. 

I also moved the analytics calls for onNextStep, onPrint to their respective function and removed the `.bind(this)` from their FormButtons.

noxee commented:
I've update the save to disk button to now use an icon instead of text.

timmyc commented:
@noxee thanks for the updates - the code is looking good!  Just tested out the branch again, and am experiencing some oddness when I play with the viewport width.  First issue is encountered prior to hitting first "tablet sized" breakpoint:

![two-step_authentication_ _wordpress_com](https://cloud.githubusercontent.com/assets/22080/14965609/2deaab68-1063-11e6-9003-10e2e4fd0dba.png)

On Mobile things look better, but still kind of funky looking to me for some reason:

![two-step_authentication_ _wordpress_com 2](https://cloud.githubusercontent.com/assets/22080/14965622/427f711c-1063-11e6-81b6-c6ea7a27099a.png)

One item that is still missing is the tooltip on the download button.

I'm going to flag design review here in hopes a designer can chime in on the above screenshots with ideas

lancewillett commented:
Reviewing old pull requests: this and related #4603 look very close to merge-able and are good improvements. Can we get these in soon?

Ping @timmyc for code quality and @drw158 for design review.

drw158 commented:
Hm not sure how this relates to #4603 seems to more up to date as far as the design goes. We should look to that issue for the design.

noxee commented:
I'm going to close this PR as all the changes from this branch were merged into #4603 as it was becoming cumbersome to keep the two separate.

Sorry for not updating this sooner.
